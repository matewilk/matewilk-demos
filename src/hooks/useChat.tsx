import { useState, useEffect } from "react";
import { gql, useSubscription, useMutation } from "@apollo/client";

export const useChat = ({ chatId }: { chatId: string }) => {
  const [messages, setMessages] = useState([] as any[]);

  const CHAT_SUBSCRIPTION = gql`
    subscription Subscription($chatId: String) {
      chat(id: $chatId) {
        id
        text
        userId
        userName
      }
    }
  `;

  // TODO: handle subscription errors
  // https://www.apollographql.com/docs/react/data/subscriptions/#handling-errors
  // also usefull when deving the subscription and schema changes
  const subscription = useSubscription(CHAT_SUBSCRIPTION, {
    variables: { chatId },
  });
  const { data } = subscription;

  useEffect(() => {
    if (data) {
      setMessages((prev) => [
        ...prev,
        {
          id: data.chat.id,
          text: data.chat.text,
          userId: data.chat.userId,
          userName: data.chat.userName,
        },
      ]);
    }
  }, [data]);

  const SEND_MESSAGE = gql`
    mutation Mutation(
      $text: String!
      $chatId: String!
      $userId: String!
      $userName: String!
    ) {
      sendMessage(
        text: $text
        chatId: $chatId
        userId: $userId
        userName: $userName
      ) {
        id
        text
        userId
        userName
      }
    }
  `;

  const mutation = useMutation(SEND_MESSAGE);

  return {
    messages,
    subscription,
    mutation,
  };
};
