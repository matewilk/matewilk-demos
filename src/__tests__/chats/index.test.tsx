import { render } from "@/utils/test-utils";
import { ApolloProvider } from "@apollo/client";
import { setupServer } from "msw/node";
import { rest } from "msw";

import { apolloClient } from "@/hooks/useWebSocketClient";
import Chats, { getServerSideProps } from "@/pages/chats/index";

// eslint-disable-next-line react/display-name
jest.mock("@/components/layout/Header", () => () => <div>Header</div>);

const expectedChats = [
  {
    id: "1",
    name: "Chat 1",
  },
  {
    id: "2",
    name: "Chat 2",
  },
];

const haqndlers = [
  rest.get("http://localhost:3000/api/chats", (req, res, ctx) => {
    return res(ctx.json(expectedChats));
  }),
];

const server = setupServer(...haqndlers);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

describe("Chats", () => {
  it("should render a list of chats and header", () => {
    const { container } = render(
      <ApolloProvider client={apolloClient()}>
        <Chats chats={expectedChats} />
      </ApolloProvider>
    );
    expect(container).toMatchSnapshot();
  });
});

describe("serverSideProps", () => {
  it("should fetch chats", async () => {
    const contextMock = {};
    const res = await getServerSideProps(contextMock as any);

    const expected = { props: { chats: expectedChats } };
    expect(res).toEqual(expected);
  });
});
