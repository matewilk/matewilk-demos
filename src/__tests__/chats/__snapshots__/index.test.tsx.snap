// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chats should render a list of chats and header 1`] = `
<div>
  <div
    class="flex min-h-screen flex-col bg-gradient-to-t from-blue-400 to-[#2e026d]"
  >
    <div>
      Header
    </div>
    <section
      aria-label="chat"
      class="h-1 flex-grow bg-black/10 font-thin text-slate-300"
      id="chat"
    >
      <div
        class="mx-auto h-full max-w-3xl py-10"
      >
        <div
          class="frosted flex h-full overflow-auto rounded-xl"
        >
          <ul
            class="flex w-full flex-col gap-2 py-5 px-5"
          >
            <li
              class="frosted w-full cursor-pointer rounded-md p-5"
            >
              Chat 1
            </li>
            <li
              class="frosted w-full cursor-pointer rounded-md p-5"
            >
              Chat 2
            </li>
          </ul>
        </div>
      </div>
    </section>
  </div>
</div>
`;
